src/Common/src/System/Collections/Generic/EnumerableHelpers.cs

```diff
@@ -2,23 +2,220 @@
+            int read = first.Length; // number of items we've read so far, updated every time we exhaust a buffer
+
+            while (true)
+            {
+                int index = 0; // index into the current buffer
+
+                // Continue reading the data into the current buffer
+                do
+                {
+                    current[index++] = en.Current;
+
+                    if (index == current.Length)
+                    {
+                        index = -1;
+                        break;
+                    }
+                }
+                while (en.MoveNext());
+
+                // If the index is -1, we called Current but not MoveNext, however
+                // the loop expects to be entered with MoveNext already having been
+                // called. So we need to call MoveNext ourselves in that case.
+                if (index >= 0 || !en.MoveNext())
+                {
+                    // We've finished enumerating the sequence.
+                    // Copy the data from the first buffer, then walk the list of buffers
+                    // and copy the data from those. Finally, copy the data from the
+                    // buffer we were just processing.
+                    
+                    int remainder = index >= 0 ? index : current.Length; // If we got here from !en.MoveNext() that means index == -1 and there was just enough space
+                    int finalLength = read + remainder;
+
+                    var result = new T[finalLength];
+
+                    // Use Array.Copy for everything, since this codepath is only taken for
+                    // buffers of length > 32
+                    Array.Copy(first, 0, result, 0, first.Length);
+
+                    // Copy from the buffers in the list that came before this one
+                    int copied = first.Length;
+                    for (int i = 0; i < buffers.Count; i++)
+                    {
+                        T[] previous = buffers[i];
+                        Array.Copy(previous, 0, result, copied, previous.Length);
+                        copied += previous.Length;
+                    }
+
+                    // Copy the remaining data from this buffer
+                    Debug.Assert(remainder <= current.Length);
+                    Debug.Assert(copied + remainder == result.Length);
+                    Array.Copy(current, 0, result, copied, remainder);
+
+                    // Done!
+                    return result;
+                }
+
+                // Since we exhausted the buffer, update read
+                read += current.Length;
+
+                // There was not enough space in the current buffer- add it to the list,
+                // and allocate a new buffer twice our size for the next iteration.
+                buffers.Add(current);
+                int nextCapacity = current.Length * 2;
```

Comment:
@stephentoub No, since if the enumerable is between 2^30 and 2^31 - 1 elements we will not attempt to allocate an array over MaxArrayLength. Will change.